modules/openthread/platform/crypto_psa_key_allocator.c

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2025 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "crypto_psa_key_allocator.h"
+
+#ifdef CONFIG_OPENTHREAD_PSA_NVM_BACKEND_KMU
+#include <cracen_psa_kmu.h>
+
+#define OPENTHREAD_KMU_SINGLE_KEY_SLOT_SIZE (2)
+/* Reserved for six 128-bit keys, and one 256-bit key */
+#define OPENTHREAD_KMU_DEDICATED_SLOTS	    (8)
+#endif /* CONFIG_OPENTHREAD_PSA_NVM_BACKEND_KMU */
+
+otError getKeyRef(otCryptoKeyRef *aInputKeyRef, psa_key_attributes_t *aAttributes)
+{
+	if (!aInputKeyRef) {
+		return OT_ERROR_INVALID_ARGS;
+	}
+
+#if defined(CONFIG_OPENTHREAD_PSA_NVM_BACKEND_ITS)
+	/* In this case we do not need make any adjustments to the input KeyRef or Attributes */
+	return OT_ERROR_NONE;
+
+#elif defined(CONFIG_OPENTHREAD_PSA_NVM_BACKEND_KMU)
+	/* Exit function if the key is outside the expected range dedicated to persistent lifetime.
+	 * Currently we supports only 7 keys defined in the openthread/src/core/crypto/storage.hpp
+	 * file. The first six keys are 128-bit length (1 KMU slot), and the last one is 256-bit
+	 * length (2 kmu slots). If the OPENTHREAD_KMU_DEDICATED_SLOTS is changed the new
+	 * formula for converting keys must be written, because we must take into account the shift
+	 * of two slots for ECDSA private key (7th key).
+	 */
+	if (*aInputKeyRef <= CONFIG_OPENTHREAD_PSA_ITS_NVM_OFFSET ||
+	    *aInputKeyRef > CONFIG_OPENTHREAD_PSA_ITS_NVM_OFFSET + OPENTHREAD_KMU_DEDICATED_SLOTS) {
+		return OT_ERROR_NONE;
+	}
+
+	/* Convert key to KMU slot, currently all keys except ECDSA private key are 128-bit length
+	so need one KMU slot. the ECDSA privatekey is the last one on the list, so we can simply
+	convert it one by one. Decrease it by 1 to get the correct slot as aInputKeyRef starts
+	from 1.
+	 */
+	*aInputKeyRef = PSA_KEY_HANDLE_FROM_CRACEN_KMU_SLOT(
+		CRACEN_KMU_KEY_USAGE_SCHEME_RAW,
+		CONFIG_OPENTHREAD_KMU_SLOT_START +
+			(*aInputKeyRef - CONFIG_OPENTHREAD_PSA_ITS_NVM_OFFSET) - 1);
+
+	/* Convert key attributes to meet KMU requirements */
+	if (aAttributes) {
+
+		/* Set the key location and lifetime to persistent and CRACEN KMU */
+		psa_set_key_lifetime(aAttributes, PSA_KEY_LIFETIME_FROM_PERSISTENCE_AND_LOCATION(
+							  PSA_KEY_PERSISTENCE_DEFAULT,
+							  PSA_KEY_LOCATION_CRACEN_KMU));
+
+		/* We cannot store raw data in KMU, so we need to convert it to one of the supported
+			types. Currently only HMAC and ECC key types can be exported from KMU.
+			The raw keys are supposed to be used for AES operation, so if
+		   TODO: remove this part once KMU driver supports these types.
```

Comment:
Something was cut off here. So... RAW_DATA is for AES operation? Which keys are these? Why they are RAW_DATA, maybe it's OT core issue?

Reply:
Thread spec doesn't define a type and algorithm for PSKc, and implementation is as follows: https://github.com/nrfconnect/sdk-openthread/blob/main/src/core/thread/key_manager.cpp#L619 It leaves defining it for a vendor to decide which algorithm should be used. If we use PSA ITS, we can use the RAW_DATA type, and as a Vendor Algorithm we provide (psa_algorithm_t)0; to PSA API, which can be used in ITS. When it comes to KMU, currently, it is not possible to pass the RAW_DATA key type and 0 algorithm, so we must translate to something that KMU knows. Do you think the Thread should strictly specify the key type and algorithm?